### PR TITLE
[Notifier] Added 46elks notifier bridge

### PR DESCRIPTION
**Pull request by jongotlin**
| Q             | A
| ------------- | ---
| Branch?       | 6.1
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | 
| License       | MIT
| Doc PR        | symfony/symfony-docs#16350

[46elks](https://46elks.se) is a Swedish sms provider.

### CODE REVIEW

**Comment by jongotlin** [comment]
Test is failing due to the transport is named `46elks` but classes are named `FortysixElks...`. Shall I rename the transport to `fortysix-elks` or alter `FrameworkExtensionTest` somehow?

File: src/Symfony/Component/Notifier/Bridge/FortysixElks/FortysixElksTransport.php [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+/**
+ * @author Jon Gotlin <jon@jon.se>
+ */
+class FortysixElksTransport extends AbstractTransport
+{
+    protected const HOST = 'api.46elks.com';
+
+    private string $apiUsername;
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/FortysixElks/FortysixElksTransport.php]: username would be enough I guess

File: src/Symfony/Component/Notifier/Bridge/FortysixElks/FortysixElksTransport.php [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    private string $apiPassword;
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/FortysixElks/FortysixElksTransport.php]: password would be enough I guess

File: src/Symfony/Component/Notifier/Bridge/FortysixElks/FortysixElksTransport.php [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+use Symfony\Component\Notifier\Message\SmsMessage;
+use Symfony\Component\Notifier\Transport\AbstractTransport;
+use Symfony\Contracts\EventDispatcher\EventDispatcherInterface;
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+/**
+ * @author Jon Gotlin <jon@jon.se>
+ */
+class FortysixElksTransport extends AbstractTransport
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/FortysixElks/FortysixElksTransport.php]:
```suggestion
final class FortysixElksTransport extends AbstractTransport
```

File: src/Symfony/Component/Notifier/Bridge/FortysixElks/FortysixElksTransportFactory.php [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+namespace Symfony\Component\Notifier\Bridge\FortysixElks;
+
+use Symfony\Component\Notifier\Exception\UnsupportedSchemeException;
+use Symfony\Component\Notifier\Transport\AbstractTransportFactory;
+use Symfony\Component\Notifier\Transport\Dsn;
+
+/**
+ * @author Jon Gotlin <jon@jon.se>
+ */
+class FortysixElksTransportFactory extends AbstractTransportFactory
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/FortysixElks/FortysixElksTransportFactory.php]:
```suggestion
final class FortysixElksTransportFactory extends AbstractTransportFactory
```

File: src/Symfony/Component/Notifier/Bridge/FortysixElks/LICENSE [hunk 5]
```diff
@@ -0,0 +1,19 @@
+Copyright (c) 2021 Fabien Potencier
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/FortysixElks/LICENSE]:
```suggestion
Copyright (c) 2022 Fabien Potencier
```
🤷🏼‍♂️😅

File: src/Symfony/Component/Notifier/Bridge/FortysixElks/FortysixElksTransport.php (hunk 1 above)
Comment (jongotlin) [reply src/Symfony/Component/Notifier/Bridge/FortysixElks/FortysixElksTransport.php]: I used the api-prefix to not confuse with username and password that is used to log in on the website. 46elks is also using the prefix in their api documentation.

File: src/Symfony/Component/Notifier/Bridge/FortysixElks/Tests/success-response.json [hunk 6]
```diff
@@ -0,0 +1,11 @@
+{
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/FortysixElks/Tests/success-response.json]: Please move this file to a Fixtures directory

**Comment by OskarStark** [comment]
@symfony/mergers how do we proceed with the class naming? Thanks

**Comment by Nyholm** [comment]
Awesome. Thank you. 

Could you add this package to the root composer.json's `replace` section? I would also like to see fabbot.io happy (just some CS)


About the naming... Hm.. Im not sure. If you dont mind, I suggest to rename the transport to `fortysix-elks` or similar so the test passes.

**Comment by jongotlin** [comment]
Did I understand you correctly about replace @Nyholm?

I made the test pass by ignoring this transport. I think it's a (slightly) better solution than renaming the transport to something that is not the service/company name.

**Review by Nyholm — APPROVED** [review]
I could go either way. 

I saw now that there already were some exceptions in that test. 

I’m happy with this pr

**Comment by OskarStark** [comment]
@Nyholm I am not sure about the replace section 🧐

**Comment by Nyholm** [comment]
> @Nyholm I am not sure about the replace section 🧐

Oh. I was just reading the CI output. I thought we did that for all bridges. See the [job](https://github.com/symfony/symfony/blob/6.1/.github/workflows/package-tests.yml#L85-L90)

@fancyweb Maybe we did a misstake [here](https://github.com/symfony/symfony/commit/fb3f4e403c91d985267f1e0aa8e9d5d17220619e#diff-7decc51dc40f779ca71da3591e2d1b46cacc0d1724823f5137061bc0a6297fc0R85)?

**Comment by Nyholm** [comment]
@OskarStark you are correct. It should not be in the composer's replace section. The CI job was wrong. I've updated the CI job in https://github.com/symfony/symfony/pull/44894

**Comment by OskarStark** [comment]
Thanks Tobias!

File: composer.json [hunk 7]
```diff
@@ -55,6 +55,7 @@
         "symfony/polyfill-uuid": "^1.15"
     },
     "replace": {
+        "symfony/46elks-notifier": "self.version",
```
Comment (Nyholm) [inline composer.json]:
```suggestion
```

**Comment by jongotlin** [comment]
Squashed and rebased. @Nyholm @OskarStark

File: src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/FrameworkExtensionTest.php [hunk 8]
```diff
@@ -1896,7 +1896,7 @@ public function testIfNotifierTransportsAreKnownByFrameworkExtension()
 
         $container = $this->createContainerFromFile('notifier');
 
-        foreach ((new Finder())->in(\dirname(__DIR__, 4).'/Component/Notifier/Bridge')->directories()->depth(0)->exclude('Mercure') as $bridgeDirectory) {
+        foreach ((new Finder())->in(\dirname(__DIR__, 4).'/Component/Notifier/Bridge')->directories()->depth(0)->exclude(['Mercure', 'FortysixElks']) as $bridgeDirectory) {
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/FrameworkExtensionTest.php]: why is this required?

File: src/Symfony/Component/Notifier/Bridge/FortysixElks/Tests/FortysixElksTransportTest.php [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+use Symfony\Component\Notifier\Exception\TransportException;
+use Symfony\Component\Notifier\Message\ChatMessage;
+use Symfony\Component\Notifier\Message\MessageInterface;
+use Symfony\Component\Notifier\Message\SentMessage;
+use Symfony\Component\Notifier\Message\SmsMessage;
+use Symfony\Component\Notifier\Test\TransportTestCase;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+
+final class FortysixElksTransportTest extends TransportTestCase
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Notifier/Bridge/FortysixElks/Tests/FortysixElksTransportTest.php]: we dont use `final` in tests

**Review by nicolas-grekas — COMMENTED** [review]
random comment :)

File: src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/FrameworkExtensionTest.php (hunk 8 above)
Comment (jongotlin) [reply src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/FrameworkExtensionTest.php]: Transport is named `46elks` but classes are named `FortysixElks` (classnames cannot start with numbers).

**Review by nicolas-grekas — APPROVED** [review]

**Review by OskarStark — APPROVED** [review]
What about renaming all 46 to FortySix? 

Classnames, transport names etc. ?

@fabpot what's your preference?

**Comment by nicolas-grekas** [comment]
FortySix or Fortysix
But also 46 or fortysix or forty-six in the package name (and maybe in the service name of course)

**Comment by fabpot** [comment]
If we use letters instead of numbers, that would be `FortySix` (F and S uppercased), and so `forty-six` for the package name and where we need underscore names.

In any case, we need to be consistent and only use one or the other, not both.

**Comment by jongotlin** [comment]
Renamed classes and transport. Some places still refers to 46elks but that's more about the company rather than the implementation. But some are in between and I'm not sure what to use 🤷

File: src/Symfony/Component/Notifier/Bridge/FortySixElks/README.md [hunk 10]
```diff
@@ -0,0 +1,24 @@
+46elks Notifier
+===============
+
+Provides [46elks](https://46elks.se) integration for Symfony Notifier.
+
+DSN example
+-----------
+
+```
+FORTYSIX_ELKS_DSN=46elks://API_USERNAME:API_PASSWORD@default?from=FROM
```
Comment (stof) [inline src/Symfony/Component/Notifier/Bridge/FortySixElks/README.md]: wrong scheme here. It is not the one used in the code.

File: src/Symfony/Component/Notifier/Bridge/FortySixElks/Tests/FortySixElksTransportFactoryTest.php [hunk 11]
```diff
@@ -0,0 +1,41 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+$identityNumber */
```
Comment (stof) [inline src/Symfony/Component/Notifier/Bridge/FortySixElks/Tests/FortySixElksTransportFactoryTest.php]:
```suggestion
 */
```

**Review by stof — CHANGES_REQUESTED** [review]

**Review by OskarStark — APPROVED** [review]

File: src/Symfony/Component/Notifier/Bridge/FortySixElks/FortySixElksTransportFactory.php [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+        $host = 'default' === $dsn->getHost() ? null : $dsn->getHost();
+        $from = $dsn->getRequiredOption('from');
+
+        return (new FortySixElksTransport($this->getUser($dsn), $this->getPassword($dsn), $from, $this->client, $this->dispatcher))->setHost($host)->setPort($dsn->getPort());
+    }
+
+    protected function getSupportedSchemes(): array
+    {
+        return ['fortysixelks'];
```
Comment (fabpot) [inline src/Symfony/Component/Notifier/Bridge/FortySixElks/FortySixElksTransportFactory.php]: Wouldn't it be more readable to use `forty-six-elks` instead?

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Notifier/Bridge/FortySixElks/FortySixElksTransportFactory.php (hunk 12 above)
Comment (jongotlin) [reply src/Symfony/Component/Notifier/Bridge/FortySixElks/FortySixElksTransportFactory.php]: No other transport schemas are named with dashes. Agree with your comment but named it similar to the others.

**Comment by jongotlin** [comment]
Regarding naming, isn't `fortysix-elks` more correct than `forty-six-elks`? It is referring to `46` and not `40 6`.

**Comment by stof** [comment]
@jongotlin the textual version of 46 is `forty-six`, not `fortysix`

**Comment by jongotlin** [comment]
@stof, ok thanks. Didn't know.

File: src/Symfony/Component/Notifier/Bridge/FortySixElks/FortySixElksTransportFactory.php (hunk 12 above)
Comment (fabpot) [reply src/Symfony/Component/Notifier/Bridge/FortySixElks/FortySixElksTransportFactory.php]: I've just merged the Orange SMS one with an `orange-sms` scheme. Still think this is mo readable. We could even fix the other ones by adding an additional scheme with dashes for even more consistency (could be a follow-up PR).

Comment (jongotlin) [reply src/Symfony/Component/Notifier/Bridge/FortySixElks/FortySixElksTransportFactory.php]: Agreed! Renamed.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @jongotlin.
